### PR TITLE
[SPARK-16421][EXAMPLES][ML]  Improve ML Example Outputs

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

Improve example outputs to better reflect the functionality that is being presented.  This mostly consisted of modifying what was printed at the end of the example, such as calling show() with truncate=False, but sometimes required minor tweaks in the example data to get relevant output.  Explicitly set parameters when they are used as part of the example.  Fixed Java examples that failed to run because of using old-style MLlib Vectors or problem with schema.  Synced examples between different APIs.
## How was this patch tested?

Ran each example for Scala, Python, and Java and made sure output was legible on a terminal of width 100.

### CODE REVIEW

**Comment by BryanCutler** [comment]
ping @mengxr @jkbradley @MLnick , any of you mind taking a look at this?  There were a few Java examples I fixed up that wouldn't run because of using mllib.linalg.Vectors.  If it would be easier, I could separate those in another PR to get that in asap.  Thanks!

**Comment by MLnick** [comment]
@BryanCutler yeah if there are some changes that are more bug-fixes to make the examples work, let's separate those out into a new JIRA & PR. That should be a little higher priority for `2.0.1`

File: examples/src/main/java/org/apache/spark/examples/ml/JavaIsotonicRegressionExample.java [hunk 1]
```diff
@@ -50,8 +50,8 @@ public static void main(String[] args) {
     IsotonicRegression ir = new IsotonicRegression();
     IsotonicRegressionModel model = ir.fit(dataset);
 
-    System.out.println("Boundaries in increasing order: " + model.boundaries());
-    System.out.println("Predictions associated with the boundaries: " + model.predictions());
+    System.out.println("Boundaries in increasing order: " + model.boundaries() + "\n");
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaIsotonicRegressionExample.java]: No big deal, but why the extra line break?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaGaussianMixtureExample.java [hunk 2]
```diff
@@ -54,6 +54,7 @@ public static void main(String[] args) {
 
     // Output the parameters of the mixture model
     for (int i = 0; i < model.getK(); i++) {
+      System.out.println("Gaussian " + i);
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaGaussianMixtureExample.java]: Why split over two statements?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaMaxAbsScalerExample.java [hunk 3]
```diff
@@ -34,10 +44,17 @@ public static void main(String[] args) {
       .getOrCreate();
 
     // $example on$
-    Dataset<Row> dataFrame = spark
-      .read()
-      .format("libsvm")
-      .load("data/mllib/sample_libsvm_data.txt");
+    List<Row> data = Arrays.asList(
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaMaxAbsScalerExample.java]: Does the data change here? why change from reading the file?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaNGramExample.java [hunk 4]
```diff
@@ -55,16 +55,12 @@ public static void main(String[] args) {
 
     Dataset<Row> wordDataFrame = spark.createDataFrame(data, schema);
 
-    NGram ngramTransformer = new NGram().setInputCol("words").setOutputCol("ngrams");
+    NGram ngramTransformer = new NGram().setN(2).setInputCol("words").setOutputCol("bigrams");
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaNGramExample.java]: I suppose this doesn't hurt, but ngrams was still fairly OK

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public static void main(String[] args) {
 
     // compute the classification error on test data.
     double accuracy = evaluator.evaluate(predictions);
-    System.out.println("Test Error : " + (1 - accuracy));
+    System.out.println("Test Error = " + (1 - accuracy));
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Some of these changes feel kind of trivial, but I guess this is for consistency. But other new System.out.println statements use : not =

File: examples/src/main/java/org/apache/spark/examples/ml/JavaPolynomialExpansionExample.java [hunk 6]
```diff
@@ -48,23 +48,19 @@ public static void main(String[] args) {
       .setDegree(3);
 
     List<Row> data = Arrays.asList(
-      RowFactory.create(Vectors.dense(-2.0, 2.3)),
+      RowFactory.create(Vectors.dense(2.0, 1.0)),
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaPolynomialExpansionExample.java]: Why this change?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaTokenizerExample.java [hunk 7]
```diff
@@ -23,17 +23,23 @@
...
 import org.apache.spark.ml.feature.Tokenizer;
+import org.apache.spark.sql.api.java.UDF1;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Row;
 import org.apache.spark.sql.RowFactory;
-import org.apache.spark.sql.types.DataTypes;
-import org.apache.spark.sql.types.Metadata;
-import org.apache.spark.sql.types.StructField;
-import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.*;
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaTokenizerExample.java]: Here imports are collapsed to *; elsewhere a * import is expanded. I might generally not touch these, but, the standard is usually to avoid wildcard imports by default

File: examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java [hunk 8]
```diff
@@ -55,10 +56,14 @@ public static void main(String[] args) {
...
       .setVectorSize(3)
       .setMinCount(0);
+
     Word2VecModel model = word2Vec.fit(documentDF);
     Dataset<Row> result = model.transform(documentDF);
-    for (Row r : result.select("result").takeAsList(3)) {
-      System.out.println(r);
+
+    for (Row row : result.collectAsList()) {
+      java.util.List text = row.getList(0);
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java]: Import List

**Comment by srowen** [comment]
It's probably OK on the whole, improving or standardizing examples slightly. I left a number of small questions. Some of the changes didn't feel quite worth making but maybe I miss the logic.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaIsotonicRegressionExample.java (hunk 1 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaIsotonicRegressionExample.java]: The 2 arrays that are printed are large and all the output get clumped together, looking like a huge block of text, so adding some separation makes it a bit more readable.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaGaussianMixtureExample.java (hunk 2 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaGaussianMixtureExample.java]: Yeah the 2 print statements could be combined.  I was probably just trying not to cram too much together, but I think it would be fine.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaMaxAbsScalerExample.java (hunk 3 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaMaxAbsScalerExample.java]: The data in the file is fine, but uses sparse vectors so that when the result is output, it doesn't really show anything.  Using just a small sample dataset, you can see what it is doing from the output

before
```
+-----+--------------------+--------------------+
|label|            features|      scaledFeatures|
+-----+--------------------+--------------------+
|  0.0|(692,[127,128,129...|(692,[127,128,129...|
|  1.0|(692,[158,159,160...|(692,[158,159,160...|
|  1.0|(692,[124,125,126...|(692,[124,125,126...|
```
after
```
+--------------+----------------+                                               
|      features|  scaledFeatures|
+--------------+----------------+
|[1.0,0.1,-8.0]|[0.25,0.01,-1.0]|
|[2.0,1.0,-4.0]|  [0.5,0.1,-0.5]|
|[4.0,10.0,8.0]|   [1.0,1.0,1.0]|
```

File: examples/src/main/java/org/apache/spark/examples/ml/JavaNGramExample.java (hunk 4 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaNGramExample.java]: I really only think that the param `N` should be set explicitly.  Looking back, changing the column name was not necessary, let me change that back.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java (hunk 5 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Yeah, I was just trying to make things like this consistent with other similar examples.  I think I just saw "=" used more often, but it really doesn't make a difference to me.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaPolynomialExpansionExample.java (hunk 6 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaPolynomialExpansionExample.java]: The fractional part makes the output a little ugly, where as using whole numbers is more readable and still shows the transform

before
```
[[-2.0,4.0,-8.0,2.3,-4.6,9.2,5.289999999999999,-10.579999999999998,12.166999999999996]]
[[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]]
[[0.6,0.36,0.216,-1.1,-0.66,-0.396,1.2100000000000002,0.7260000000000001,-1.3310000000000004]]
```

after
```
+----------+------------------------------------------+
|features  |polyFeatures                              |
+----------+------------------------------------------+
|[2.0,1.0] |[2.0,4.0,8.0,1.0,2.0,4.0,1.0,2.0,1.0]     |
|[0.0,0.0] |[0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]     |
|[3.0,-1.0]|[3.0,9.0,27.0,-1.0,-3.0,-9.0,1.0,3.0,-1.0]|
```

File: examples/src/main/java/org/apache/spark/examples/ml/JavaTokenizerExample.java (hunk 7 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaTokenizerExample.java]: I agree that wildcards should be avoided, not sure what happened here.  It might have been an automatic thing from the IDE, I'll revert this.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java (hunk 8 above)
Comment (BryanCutler) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java]: List was already imported, but this should be `List text = ...`

**Comment by BryanCutler** [comment]
> @BryanCutler yeah if there are some changes that are more bug-fixes to make the examples work, let's separate those out into a new JIRA & PR. That should be a little higher priority for 2.0.1

Sure @MLnick , I realized I should probably do that about half-way into this.  I'll make another JIRA and fix the Java errors there.

**Comment by BryanCutler** [comment]
Thanks for the review @srowen!  I added some before/after outputs, so hopefully some of the changes make more sense.  I'll fix up the rest after I make another JIRA for the Java errors.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java (hunk 8 above)
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaWord2VecExample.java]: Yeah, just saying it's also fully qualified here. It could have a generic bound too.

**Comment by BryanCutler** [comment]
This has been updated since fixing the errors in Java @srowen @MLnick .  I know most of these changes are trivial, but will hopefully make some of the examples easier to follow.  Thanks!

**Comment by srowen** [comment]
There's a lot of change here; I skimmed it and it all looks generally positive, adding some consistency or clarification, or a fix in some cases. Is sample_libsvm_data.txt used anymore then? it's low risk to merge because they're example changes. I'm OK with it.

**Comment by BryanCutler** [comment]
Thanks for taking another look @srowen.  `sample_libsvm_data.txt` is still used but it looks these are 
never referenced 

```
sample_tree_data.csv
pagerank_data.txt
lr_data.txt
random.data
```
I can't place where `sample_tree_data.csv` might have belonged, `pagerank_data.txt` is obvious (just missing reference in usage), and `lr_data.txt`/`random.data` look like labeled points probably from some older MLlib examples.

**Comment by BryanCutler** [comment]
attaching a quick audit of example data files and what examples reference them, take from this branch
[spark_example_data_audit.txt](https://github.com/apache/spark/files/402881/spark_example_data_audit.txt)

**Comment by srowen** [comment]
I think it's fine to remove files that aren't referenced here too.

**Comment by BryanCutler** [comment]
Ok, I removed these data files
```
sample_tree_data.csv
lr_data.txt
random.data
```
and added example usage to reference `pagerank_data.txt`

**Comment by srowen** [comment]
Merged to master

**Comment by BryanCutler** [comment]
Thanks @srowen!
